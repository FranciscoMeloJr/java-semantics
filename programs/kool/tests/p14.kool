class a {
  var i:int, j:int;
  method a() {
    i = 0;
    j = ++i+1;
  }
  method aa():int {
    j = ++i+1;
    return j;
  }
  method f():int {
    return aa();
  }
  method g():int {
    return f();
  }
  method h():int {
    return i+j;
  }
}

class b extends a {
  var j:int, k:int;
  method aa():int {
    return bb();
  }
  method b() {
    super.a();
    j = 10;
    k = j+1;
  }
  method bb():int {
    k = ++j+1;
    return k;
  }
  method g():int {
    return super.h();
  }
  method h():int {
    return g();
  }
}

class c extends b {
  method aa():int {
    return super.aa();
  }
  method bb():int {
    return super.bb();
  }
  method c() {
    i = 100;
    j = i+1;
    k = j+1;
  }
  method g():int {
    return i+k*j;
  }
}

class main {
  method p(o:class a) {
    print(o.f());print(o.g());print(o.h());
  }
  method main() {
    p(new a());
    p(new b());
    p(new c());
  }
}

