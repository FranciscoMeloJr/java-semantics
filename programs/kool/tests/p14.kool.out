34712331031060610606
